core/network/impl/protocols/state_protocol_impl.cpp

```diff
@@ -0,0 +1,339 @@
+                   peer_id.value());
+          self->onIncomingStream(std::forward<decltype(stream)>(stream));
+          return;
+        }
+        self->log_->warn("Handled {} protocol stream from unknown peer",
+                         self->protocol_);
+      }
+    });
+    return true;
+  }
+
+  bool StateProtocolImpl::stop() {
+    return true;
+  }
+
+  void StateProtocolImpl::onIncomingStream(std::shared_ptr<Stream> stream) {
+    BOOST_ASSERT(stream->remotePeerId().has_value());
+
+    readRequest(stream);
+  }
+
+  void StateProtocolImpl::newOutgoingStream(
+      const PeerInfo &peer_info,
+      std::function<void(outcome::result<std::shared_ptr<Stream>>)> &&cb) {
+    SL_DEBUG(log_, "Connect for {} stream with {}", protocol_, peer_info.id);
+
+    host_.newStream(
+        peer_info.id,
+        protocol_,
+        [wp = weak_from_this(), peer_id = peer_info.id, cb = std::move(cb)](
+            auto &&stream_res) mutable {
+          auto self = wp.lock();
+          if (not self) {
+            cb(ProtocolError::GONE);
+            return;
+          }
+
+          if (not stream_res.has_value()) {
+            SL_VERBOSE(
+                self->log_,
+                "Error happened while connection over {} stream with {}: {}",
+                self->protocol_,
+                peer_id,
+                stream_res.error().message());
+            cb(stream_res.as_failure());
+            return;
+          }
+          auto &stream = stream_res.value();
+
+          SL_DEBUG(self->log_,
+                   "Established connection over {} stream with {}",
+                   self->protocol_,
+                   peer_id);
+
+          cb(std::move(stream));
+        });
+  }
+
+  void StateProtocolImpl::readRequest(std::shared_ptr<Stream> stream) {
+    auto read_writer = std::make_shared<ProtobufMessageReadWriter>(stream);
+
+    SL_DEBUG(log_,
+             "Read request from incoming {} stream with {}",
+             protocol_,
+             stream->remotePeerId().value());
+
+    read_writer->read<StateRequest>([stream, wp = weak_from_this()](
+                                        auto &&state_request_res) mutable {
+      auto self = wp.lock();
+      if (not self) {
+        stream->reset();
+        return;
+      }
+
+      if (not state_request_res.has_value()) {
+        SL_VERBOSE(self->log_,
+                   "Error at read request from incoming {} stream with {}: {}",
+                   self->protocol_,
+                   stream->remotePeerId().value(),
+                   state_request_res.error().message());
+
+        stream->reset();
+        return;
+      }
+
+      auto &state_request = state_request_res.value();
+
+      if (self->log_->level() >= log::Level::VERBOSE) {
+        std::string keys;
+        if (!state_request.start.empty()) {
+          keys = " starting with keys [";
+          for (const auto &key : state_request.start) {
+            keys += key.toHex() + ",";
```

Comment:
remove possible reallocation for temporary object
```c++
keys += key.toHex();
keys += ',';

Reply:
Done